### PR TITLE
[SPARK-37983][SQL] Back out agg build time metrics from sort aggregate

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

This is a followup of https://issues.apache.org/jira/browse/SPARK-37564 . I realize the agg build time metrics for sort aggregate is actually not correctly recorded. We don't have a hash build phase for sort aggregate, so there is really no way to measure so-called build time for sort aggregate. So here I make the change to back out the change introduced in https://github.com/apache/spark/pull/34826 for agg build time metric.

### Why are the changes needed?

To not report confusing SQL metrics `aggTime` for users.

### Does this PR introduce _any_ user-facing change?

No, because https://github.com/apache/spark/pull/34826 is not released yet.

### How was this patch tested?

The existing unit test introduced in https://github.com/apache/spark/pull/34826 .

### CODE REVIEW

**Comment by c21** [comment]
@cloud-fan could you help take a look when you have time? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala [hunk 1]
```diff
@@ -44,8 +42,7 @@ case class SortAggregateExec(
   with AliasAwareOutputOrdering {
 
   override lazy val metrics = Map(
-    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
-    "aggTime" -> SQLMetrics.createTimingMetric(sparkContext, "time in aggregation build"))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala:48]: unrelated to this PR: the name `aggTime` is a bit misleading. Is it actually the time to build the hash table?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala:48]: @cloud-fan - for hash agg and object hash agg, yes, the `aggTime` is the time to build the hash table. For sort aggregate, there's no proper definition of agg time, as the whole time during sort aggregate is doing aggregate. There's not that useful to measure the execution time for sort aggregate.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala:48]: if there is no group col, does `HashAggregate` always report `aggTime` as 0?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala:48]: > if there is no group col, does HashAggregate always report aggTime as 0?

No, `aggTime` is the time for `HashAggregate` to aggregate all input rows.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortAggregateExec.scala:48]: @cloud-fan - yeah, I agree with you, that `aggTime` is misleading. It's the time to build hash table when group columns exist.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregateCodegenSupport.scala [hunk 2]
```diff
@@ -47,6 +47,12 @@ trait AggregateCodegenSupport
    */
   private var bufVars: Seq[Seq[ExprCode]] = _
 
+  /**
+   * Whether this operator has an aggregate build phase.
+   * Only [[HashAggregateExec]] has build phase now.
+   */
+  protected def hasAggBuild: Boolean
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregateCodegenSupport.scala:54]: shall we rename it to `needHashTable`? And it should be false in `HashAggregateExec` if group by col is empty.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregateCodegenSupport.scala:54]: do we want to report `aggTime` as 0 now for `HashAggregate` without group by columns?

I am fine with changing to report 0 now, as the SQL metrics shown up as `time in aggregation build` in Spark UI for end users. There's no build for `HashAggregate` without group by columns, so I think it's a valid fix. Just want to mention here as this is a user-facing change.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregateCodegenSupport.scala:54]: ok maybe we can do the change in a new PR. Let's just use a proper name in this PR.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregateCodegenSupport.scala]: @cloud-fan - sure, changed to `needHashTable`. Leave `HashAggregateExec.needHashTable` to be true, to maintain existing behavior for now.

**Comment by c21** [comment]
@cloud-fan - the PR is ready for review again, thanks.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**